gobblin-yarn/src/main/java/gobblin/yarn/GobblinYarnAppLauncher.java

```diff
@@ -0,0 +1,598 @@
+    int maximumMemoryCapacity = newApplicationResponse.getMaximumResourceCapability().getMemory();
+    if (memoryMbs > maximumMemoryCapacity) {
+      LOGGER.info(String.format("Specified AM memory [%d] is above the maximum memory capacity [%d] of the "
+          + "cluster, using the maximum memory capacity instead.", memoryMbs, maximumMemoryCapacity));
+      memoryMbs = maximumMemoryCapacity;
+    }
+
+    int vCores = this.config.getInt(ConfigurationConstants.APP_MASTER_CORES_KEY);
+    int maximumVirtualCoreCapacity = newApplicationResponse.getMaximumResourceCapability().getVirtualCores();
+    if (vCores > maximumVirtualCoreCapacity) {
+      LOGGER.info(String.format("Specified AM vcores [%d] is above the maximum vcore capacity [%d] of the "
+          + "cluster, using the maximum vcore capacity instead.", memoryMbs, maximumMemoryCapacity));
+      vCores = maximumVirtualCoreCapacity;
+    }
+
+    // Set up resource type requirements for ApplicationMaster
+    return Resource.newInstance(memoryMbs, vCores);
+  }
+
+  private void addAppMasterLocalResources(ApplicationId applicationId,
+      Map<String, LocalResource> resources) throws IOException {
+    Path appWorkDir = YarnHelixUtils.getAppWorkDirPath(this.fs, this.appName, applicationId);
+    Path appMasterWorkDir = new Path(appWorkDir, ConfigurationConstants.APP_MASTER_WORK_DIR_NAME);
+
+    if (this.config.hasPath(ConfigurationConstants.LIB_JARS_DIR_KEY)) {
+      Path libJarsDestDir = new Path(appWorkDir, ConfigurationConstants.LIB_JARS_DIR_NAME);
+      addLibJars(new Path(this.config.getString(ConfigurationConstants.LIB_JARS_DIR_KEY)),
+          Optional.of(resources), libJarsDestDir);
+    }
+    if (this.config.hasPath(ConfigurationConstants.APP_MASTER_JARS_KEY)) {
+      Path appJarsDestDir = new Path(appMasterWorkDir, ConfigurationConstants.APP_JARS_DIR_NAME);
+      addAppJars(this.config.getString(ConfigurationConstants.APP_MASTER_JARS_KEY),
+          Optional.of(resources), appJarsDestDir);
+    }
+    if (this.config.hasPath(ConfigurationConstants.APP_MASTER_FILES_LOCAL_KEY)) {
+      Path appFilesDestDir = new Path(appMasterWorkDir, ConfigurationConstants.APP_FILES_DIR_NAME);
+      addAppLocalFiles(this.config.getString(ConfigurationConstants.APP_MASTER_FILES_LOCAL_KEY),
+          Optional.of(resources), appFilesDestDir);
+    }
+    if (this.config.hasPath(ConfigurationConstants.APP_MASTER_FILES_REMOTE_KEY)) {
+      addAppRemoteFiles(this.config.getString(ConfigurationConstants.APP_MASTER_FILES_REMOTE_KEY), resources);
+    }
+    if (this.config.hasPath(ConfigurationConstants.JOB_CONF_PACKAGE_PATH_KEY)) {
+      Path appFilesDestDir = new Path(appMasterWorkDir, ConfigurationConstants.APP_FILES_DIR_NAME);
+      addJobConfPackage(this.config.getString(ConfigurationConstants.JOB_CONF_PACKAGE_PATH_KEY),
+          appFilesDestDir, resources);
+    }
+  }
+
+  private void addContainerLocalResources(ApplicationId applicationId) throws IOException {
```

Comment:
This method needs to add all the app-level dependencies added by the AM, which is already on HDFS instead of local disk.